File: packages/contracts/src/forum-post.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

/**
 * The threaded "Forum Warga" board.
 *
 * This lives alongside the chat channels in `forum.ts` rather than replacing
 * them: a post has a title, a category, likes, and replies, none of which a
 * chat message has. The Forum tab shows both, as two modes.
 */
export const forumPostCategorySchema = z.enum([
  "QUESTION",
  "SUGGESTION",
  "INFORMATION",
  "ENVIRONMENT",
  "ACTIVITY",
]);

export type ForumPostCategory = z.infer<typeof forumPostCategorySchema>;

export const FORUM_POST_CATEGORY_LABELS: Record<ForumPostCategory, string> = {
  QUESTION: "Tanya Warga",
  SUGGESTION: "Usulan",
  INFORMATION: "Informasi",
  ENVIRONMENT: "Info Lingkungan",
  ACTIVITY: "Ide & Kegiatan",
};

/** `answered` keeps only posts that already have at least one reply, so a
 * warga can skip past the questions nobody has picked up yet. */
export const forumPostSortSchema = z.enum(["latest", "popular", "answered"]);

export type ForumPostSort = z.infer<typeof forumPostSortSchema>;

export const FORUM_POST_SORT_LABELS: Record<ForumPostSort, string> = {
  latest: "Terbaru",
  popular: "Populer",
  answered: "Terjawab",
};

export const forumPostListQuerySchema = z.object({
  sort: forumPostSortSchema.default("latest"),
  category: forumPostCategorySchema.optional(),
  limit: z.coerce.number().int().min(1).max(50).default(20),
});

export type ForumPostListQuery = z.infer<typeof forumPostListQuerySchema>;

const authoredFields = {
  authorUserId: z.string().uuid(),
  authorName: z.string().min(1),
  createdAt: z.string().datetime({ offset: true }),
  editedAt: z.string().datetime({ offset: true }).nullable(),
  likeCount: z.number().int().min(0),
  /** Whether the viewer has liked it — the heart is per-person, so this cannot
   * be derived from `likeCount`. */
  likedByMe: z.boolean(),
} as const;

export const forumPostSummarySchema = z.object({
  id: z.string().uuid(),
  category: forumPostCategorySchema,
  title: z.string().min(1),
  /** First lines of the body, for the card on the board. */
  excerpt: z.string(),
  imageUrls: z.array(z.string().url()),
  replyCount: z.number().int().min(0),
  ...authoredFields,
});

export type ForumPostSummary = z.infer<typeof forumPostSummarySchema>;

export const forumPostReplySchema = z.object({
  id: z.string().uuid(),
  postId: z.string().uuid(),
  body: z.string().min(1),
  replyToReplyId: z.string().uuid().nullable(),
  /** Denormalised quote of the reply being answered; null when it was
   * deleted, so the thread never resurrects removed content. */
  replyToAuthorName: z.string().nullable(),
  replyToBody: z.string().nullable(),
  ...authoredFields,
});

export type ForumPostReply = z.infer<typeof forumPostReplySchema>;

export const forumPostDetailSchema = forumPostSummarySchema.extend({
  body: z.string().min(1),
  replies: z.array(forumPostReplySchema),
});

export type ForumPostDetail = z.infer<typeof forumPostDetailSchema>;

export const forumPostListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(forumPostSummarySchema) }),
);

export type ForumPostListResponse = z.infer<typeof forumPostListResponseSchema>;

export const forumPostDetailResponseSchema = dataEnvelopeSchema(
  z.object({ post: forumPostDetailSchema }),
);

export type ForumPostDetailResponse = z.infer<typeof forumPostDetailResponseSchema>;

export const forumPostMutationResponseSchema = dataEnvelopeSchema(
  z.object({ post: forumPostSummarySchema }),
);

export type ForumPostMutationResponse = z.infer<typeof forumPostMutationResponseSchema>;

export const forumPostReplyMutationResponseSchema = dataEnvelopeSchema(
  z.object({ reply: forumPostReplySchema }),
);

export type ForumPostReplyMutationResponse = z.infer<
  typeof forumPostReplyMutationResponseSchema
>;

export const createForumPostInputSchema = z
  .object({
    category: forumPostCategorySchema.default("INFORMATION"),
    title: z.string().trim().min(5).max(240),
    body: z.string().trim().min(10).max(5_000),
    imageUrls: z.array(z.string().url()).max(5).default([]),
  })
  .strict();

export type CreateForumPostInput = z.infer<typeof createForumPostInputSchema>;

export const updateForumPostInputSchema = z
  .object({
    category: forumPostCategorySchema.optional(),
    title: z.string().trim().min(5).max(240).optional(),
    body: z.string().trim().min(10).max(5_000).optional(),
    imageUrls: z.array(z.string().url()).max(5).optional(),
  })
  .strict()
  .refine((input) => Object.keys(input).length > 0, {
    message: "Kirim setidaknya satu perubahan.",
  });

export type UpdateForumPostInput = z.infer<typeof updateForumPostInputSchema>;

export const createForumPostReplyInputSchema = z
  .object({
    body: z.string().trim().min(1).max(2_000),
    replyToReplyId: z.string().uuid().optional(),
  })
  .strict();

export type CreateForumPostReplyInput = z.infer<typeof createForumPostReplyInputSchema>;

export const updateForumPostReplyInputSchema = z
  .object({ body: z.string().trim().min(1).max(2_000) })
  .strict();

export type UpdateForumPostReplyInput = z.infer<typeof updateForumPostReplyInputSchema>;

export const forumLikeResponseSchema = dataEnvelopeSchema(
  z.object({
    likeCount: z.number().int().min(0),
    likedByMe: z.boolean(),
  }),
);

export type ForumLikeResponse = z.infer<typeof forumLikeResponseSchema>;

export const forumPostDeleteResponseSchema = dataEnvelopeSchema(
  z.object({ postId: z.string().uuid() }),
);

export type ForumPostDeleteResponse = z.infer<typeof forumPostDeleteResponseSchema>;

export const forumReplyDeleteResponseSchema = dataEnvelopeSchema(
  z.object({ replyId: z.string().uuid() }),
);

export type ForumReplyDeleteResponse = z.infer<typeof forumReplyDeleteResponseSchema>;

File: apps/mobile/lib/features/forum/domain/forum_post.dart
/// What a discussion-board post is for. Chat channels have no equivalent —
/// this only applies to the threaded board.
enum ForumPostCategory { question, suggestion, information, environment, activity }

/// `answered` keeps only posts that already have at least one reply, so a
/// warga can skip past the questions nobody has picked up yet.
enum ForumPostSort { latest, popular, answered }

const forumPostCategoryLabels = {
  ForumPostCategory.question: 'Tanya Warga',
  ForumPostCategory.suggestion: 'Usulan',
  ForumPostCategory.information: 'Informasi',
  ForumPostCategory.environment: 'Info Lingkungan',
  ForumPostCategory.activity: 'Ide & Kegiatan',
};

const forumPostSortLabels = {
  ForumPostSort.latest: 'Terbaru',
  ForumPostSort.popular: 'Populer',
  ForumPostSort.answered: 'Terjawab',
};

extension ForumPostCategoryApi on ForumPostCategory {
  String get apiValue => switch (this) {
        ForumPostCategory.question => 'QUESTION',
        ForumPostCategory.suggestion => 'SUGGESTION',
        ForumPostCategory.information => 'INFORMATION',
        ForumPostCategory.environment => 'ENVIRONMENT',
        ForumPostCategory.activity => 'ACTIVITY',
      };
}

extension ForumPostSortApi on ForumPostSort {
  String get apiValue => name;
}

ForumPostCategory forumPostCategoryFromApi(Object? value) {
  switch (value) {
    case 'QUESTION':
      return ForumPostCategory.question;
    case 'SUGGESTION':
      return ForumPostCategory.suggestion;
    case 'ENVIRONMENT':
      return ForumPostCategory.environment;
    case 'ACTIVITY':
      return ForumPostCategory.activity;
    default:
      return ForumPostCategory.information;
  }
}

class ForumPostSummary {
  const ForumPostSummary({
    required this.id,
    required this.category,
    required this.title,
    required this.excerpt,
    required this.imageUrls,
    required this.replyCount,
    required this.authorUserId,
    required this.authorName,
    required this.createdAt,
    required this.editedAt,
    required this.likeCount,
    required this.likedByMe,
  });

  final String id;
  final ForumPostCategory category;
  final String title;
  final String excerpt;
  final List<String> imageUrls;
  final int replyCount;
  final String authorUserId;
  final String authorName;
  final DateTime createdAt;
  final DateTime? editedAt;
  final int likeCount;

  /// The heart is per-person, so this cannot be derived from [likeCount].
  final bool likedByMe;

  bool get isEdited => editedAt != null;

  factory ForumPostSummary.fromJson(Map<String, dynamic> json) {
    final images = json['imageUrls'];
    final editedAt = json['editedAt'];
    return ForumPostSummary(
      id: json['id'] as String,
      category: forumPostCategoryFromApi(json['category']),
      title: json['title'] as String,
      excerpt: json['excerpt'] as String? ?? '',
      imageUrls: images is List
          ? images.map((url) => url as String).toList(growable: false)
          : const [],
      replyCount: (json['replyCount'] as num?)?.toInt() ?? 0,
      authorUserId: json['authorUserId'] as String,
      authorName: json['authorName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      editedAt: editedAt is String ? DateTime.parse(editedAt) : null,
      likeCount: (json['likeCount'] as num?)?.toInt() ?? 0,
      likedByMe: json['likedByMe'] as bool? ?? false,
    );
  }
}

class ForumPostReply {
  const ForumPostReply({
    required this.id,
    required this.postId,
    required this.body,
    required this.replyToReplyId,
    required this.replyToAuthorName,
    required this.replyToBody,
    required this.authorUserId,
    required this.authorName,
    required this.createdAt,
    required this.editedAt,
    required this.likeCount,
    required this.likedByMe,
  });

  final String id;
  final String postId;
  final String body;
  final String? replyToReplyId;

  /// Denormalised quote of the reply being answered; null when it was deleted,
  /// so the thread never resurrects removed content.
  final String? replyToAuthorName;
  final String? replyToBody;
  final String authorUserId;
  final String authorName;
  final DateTime createdAt;
  final DateTime? editedAt;
  final int likeCount;
  final bool likedByMe;

  bool get isQuote => replyToReplyId != null;
  bool get isEdited => editedAt != null;

  factory ForumPostReply.fromJson(Map<String, dynamic> json) {
    final editedAt = json['editedAt'];
    return ForumPostReply(
      id: json['id'] as String,
      postId: json['postId'] as String,
      body: json['body'] as String,
      replyToReplyId: json['replyToReplyId'] as String?,
      replyToAuthorName: json['replyToAuthorName'] as String?,
      replyToBody: json['replyToBody'] as String?,
      authorUserId: json['authorUserId'] as String,
      authorName: json['authorName'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      editedAt: editedAt is String ? DateTime.parse(editedAt) : null,
      likeCount: (json['likeCount'] as num?)?.toInt() ?? 0,
      likedByMe: json['likedByMe'] as bool? ?? false,
    );
  }
}

class ForumPostDetail {
  const ForumPostDetail({required this.post, required this.body, required this.replies});

  final ForumPostSummary post;
  final String body;
  final List<ForumPostReply> replies;

  factory ForumPostDetail.fromJson(Map<String, dynamic> json) {
    final replies = json['replies'];
    return ForumPostDetail(
      post: ForumPostSummary.fromJson(json),
      body: json['body'] as String,
      replies: replies is List
          ? replies
              .map((item) => ForumPostReply.fromJson(item as Map<String, dynamic>))
              .toList(growable: false)
          : const [],
    );
  }
}

/// "5 jam yang lalu" — the board reads as a conversation, so relative time is
/// more useful here than a timestamp.
String formatForumRelativeTime(DateTime value) {
  final elapsed = DateTime.now().difference(value);
  if (elapsed.inMinutes < 1) return 'baru saja';
  if (elapsed.inMinutes < 60) return '${elapsed.inMinutes} menit yang lalu';
  if (elapsed.inHours < 24) return '${elapsed.inHours} jam yang lalu';
  if (elapsed.inDays < 30) return '${elapsed.inDays} hari yang lalu';

  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'Mei',
    'Jun',
    'Jul',
    'Agu',
    'Sep',
    'Okt',
    'Nov',
    'Des',
  ];
  final local = value.toLocal();
  return '${local.day} ${months[local.month - 1]} ${local.year}';
}
